In [21]:
import numpy as np 
import pandas as pd 

from keras.models import Model
from keras.layers import Dense, Embedding, Input
from keras.layers import LSTM, Bidirectional, GlobalMaxPool1D, Dropout
from keras.preprocessing import text, sequence
from keras.callbacks import EarlyStopping, ModelCheckpoint

max_features = 20000
maxlen = 100


train = pd.read_csv("train.csv")
#test = pd.read_csv("test.csv")
#train = train.sample(frac=1)

list_sentences_train = train["comment_text"].fillna("nada").values
list_classes = ["toxic", "severe_toxic", "obscene", "threat", "insult", "identity_hate"]
y = train[list_classes].values
#list_sentences_test = test["comment_text"].fillna("nada").values


tokenizer = text.Tokenizer(num_words=max_features)
tokenizer.fit_on_texts(list(list_sentences_train))
list_tokenized_train = tokenizer.texts_to_sequences(list_sentences_train)
#list_tokenized_test = tokenizer.texts_to_sequences(list_sentences_test)
X_t = sequence.pad_sequences(list_tokenized_train, maxlen=maxlen)
#X_te = sequence.pad_sequences(list_tokenized_test, maxlen=maxlen)

def get_model():
    embed_size = 128
    inp = Input(shape=(maxlen, ))
    x = Embedding(max_features, embed_size)(inp)
    x = Bidirectional(LSTM(50, return_sequences=True))(x)
    x = GlobalMaxPool1D()(x)
    x = Dropout(0.1)(x)
    x = Dense(50, activation="relu")(x)
    x = Dropout(0.1)(x)
    x = Dense(6, activation="sigmoid")(x)
    model = Model(inputs=inp, outputs=x)
    model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

    return model


model = get_model()
batch_size = 32
epochs = 2


file_path="weights_base.best.hdf5"
checkpoint = ModelCheckpoint(file_path, monitor='val_loss', verbose=1, save_best_only=True, mode='min')

early = EarlyStopping(monitor="val_loss", mode="min", patience=20)


callbacks_list = [checkpoint, early] #early
model.fit(X_t, y, batch_size=batch_size, epochs=epochs, validation_split=0.1, callbacks=callbacks_list)

#model.load_weights(file_path)

#y_predict = model.predict(X_te)



#sample_submission = pd.read_csv("sample_submission.csv")

#sample_submission[list_classes] = y_predict



#sample_submission.to_csv("baseline.csv", index=False)

Train on 143613 samples, validate on 15958 samples
Epoch 1/2
143613/143613 [==============================] - 1069s 7ms/step - loss: 0.2844 - acc: 0.9907 - val_loss: 0.2791 - val_acc: 0.9939

Epoch 00001: val_loss improved from inf to 0.27913, saving model to weights_base.best.hdf5
Epoch 2/2
143613/143613 [==============================] - 1116s 8ms/step - loss: 0.2668 - acc: 0.9926 - val_loss: 0.2764 - val_acc: 0.9896

Epoch 00002: val_loss improved from 0.27913 to 0.27637, saving model to weights_base.best.hdf5


In [ ]:


#maxlens=[100,200,300]
#batch_sizes=[32]

In [7]:
train = pd.read_csv("train.csv")

In [16]:
comment_len_bin=dict()

comment_len_bin['total']=train.shape[0]
comment_len_bin['100']=0
comment_len_bin['150']=0
comment_len_bin['200']=0
comment_len_bin['250']=0
comment_len_bin['300']=0

for index,row in train.iterrows():
    
    if len(row['comment_text'].split(' '))<=100:
        comment_len_bin['100']+=(1)
        
    if len(row['comment_text'].split(' '))<=150:
        comment_len_bin['150']+=(1)

    if len(row['comment_text'].split(' '))<=200:
        comment_len_bin['200']+=(1)

    if len(row['comment_text'].split(' '))<=250:
        comment_len_bin['250']+=(1)

    if len(row['comment_text'].split(' '))<=300:
        comment_len_bin['300']+=(1)
        
    

In [14]:
comment_len_bin

{'100': 130994, '150': 0, '200': 0, '250': 0, '300': 0, 'total': 159571}

In [19]:
for i in ['100','150','200','250','300']:
    print('{:6.2f}'.format(100*comment_len_bin[i]/comment_len_bin['total']),'% of comments have length equal or less than', i, 'words')

 82.09 % of comments have length equal or less than 100 words
 89.78 % of comments have length equal or less than 150 words
 93.61 % of comments have length equal or less than 200 words
 95.75 % of comments have length equal or less than 250 words
 96.95 % of comments have length equal or less than 300 words


# Lets Automate

In [ ]:
!wget('https://s3.amazonaws.com/danicic-w266-final/train.csv')

In [1]:
import pandas as pd
import sys
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from scipy.sparse import hstack
from sklearn import model_selection
import numpy as np
import tensorflow  as tf
import keras
import h5py
from keras.models import load_model


from keras.models import Model
from keras.layers import Dense, Embedding, Input
from keras.layers import LSTM, Bidirectional, GlobalMaxPool1D, MaxPool1D, Dropout
from keras.preprocessing import text, sequence
from keras.callbacks import EarlyStopping, ModelCheckpoint

from keras.metrics import categorical_accuracy

/Users/ChrisD/anaconda3/envs/tensorflow/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [10]:
# import data
np.random.seed(412)

print('Importing data...')
train_all=pd.read_csv('train.csv')

# 80/20 split
train_80, test_20 = model_selection.train_test_split(train_all,test_size=0.2)

#preprocessing

print('Splitting data...')

#training data
x_train = train_80['comment_text'].fillna("nada").values
y_train = train_80[['toxic','severe_toxic','obscene','threat','insult','identity_hate']].values

#testing validation data (not for training model, just validation for )

x_test = test_20['comment_text'].fillna("nada").values
y_test = test_20[['toxic','severe_toxic','obscene','threat','insult','identity_hate']].values

# check lengths

print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

Importing data...
Splitting data...
127656 train sequences
31915 test sequences


## Model Parameters:

In [11]:
max_feature_list=[20000,10000]
EPOCHS=4
maxlen_list=[150,250] 
dropout_list=[0.1,0.2]

## tokenize

In [4]:
#tokenizer = text.Tokenizer(num_words=max_feature_list[0])

#tokenizer.fit_on_texts(list(train_all["comment_text"].fillna("nada").values)) #fit on all comment_text

##create tokenized comments
#list_tokenized_train = tokenizer.texts_to_sequences(x_train)  #training (80% of train.csv)
#list_tokenized_test = tokenizer.texts_to_sequences(x_test)  #testing (20% of train.csv)

*using 20000 max features*

## create LSTM model maker function

In [6]:
def get_model(maxlen,max_features,dropout,embed_size=128):
    
    embed_size=embed_size #default to 128
    maxlen=maxlen # max length of sequence input
    max_features=max_features  # max vocab
    dropout=dropout
    
    inp = Input(shape=(maxlen, ))
    
    x = Embedding(max_features, embed_size)(inp)
    
    x = Bidirectional(LSTM(50, return_sequences=False))(x)
    
    #x = GlobalMaxPool1D()(x)
    
    #x = Dropout(dropout)(x)
    
    x = Dense(50, activation="relu")(x)
    
    x = Dropout(dropout)(x)
    
    x = Dense(6, activation="sigmoid")(x)
    
    model = Model(inputs=inp, outputs=x)
    
    model.compile(loss='binary_crossentropy',optimizer='adam',metrics=[categorical_accuracy])

    return model


## Tokenize and fit model

In [24]:
max_feature_list=[20000,10000]
maxlen_list=[150,250] 
dropout_list=[0.1,0.2]
EPOCHS=4

for max_feature in [max_feature_list[1]]:
    
    print('tokenizing')
    tokenizer = text.Tokenizer(num_words=max_feature)

    tokenizer.fit_on_texts(list(train_all["comment_text"].fillna("nada").values)) #fit on all comment_text

    #create tokenized comments
    print('creating sequences')
    list_tokenized_train = tokenizer.texts_to_sequences(x_train)  #training (80% of train.csv)
    list_tokenized_test = tokenizer.texts_to_sequences(x_test)  #testing (20% of train.csv)
    
    for maxlen in [maxlen_list[1]]:
        
        for dropout in [dropout_list[0]]:
            
            print('padding maxlen=',maxlen)
            x_train_pad = sequence.pad_sequences(list_tokenized_train, maxlen=maxlen)
            x_test_pad = sequence.pad_sequences(list_tokenized_test, maxlen=maxlen)
            
            print('building model')
            model=get_model(maxlen=maxlen,max_features=max_feature,dropout=dropout,embed_size=128)
            
            batch_size = 32


            weight_file_path="weights_base_{}_{}_{}.hdf5".format(max_feature,maxlen,dropout)
            model_file_path='bidirectional_lstm_globMP_relu_sigmoid_maxfeat{}_maxlen{}_dropout{}'.format(max_feature,maxlen,dropout)

            checkpoint = ModelCheckpoint(weight_file_path, monitor='val_loss', verbose=1, save_best_only=True, mode='min')

            early = EarlyStopping(monitor="val_loss", mode="min", patience=20)

            callbacks_list = [checkpoint, early] #early
            
            #fit model
            print('fitting model with max_features={}, maxlen = {}, and dropout = {}'.format(max_feature,maxlen,dropout))
            model.fit(x_train_pad, y_train, batch_size=batch_size, epochs=EPOCHS, validation_split=0.1, callbacks=callbacks_list)
            
            #save model
            model.save(model_file_path)


tokenizing
creating sequences
padding maxlen= 250
building model
fitting model with max_features=10000, maxlen = 250, and dropout = 0.1
Train on 114890 samples, validate on 12766 samples
Epoch 1/4
114890/114890 [==============================] - 1787s 16ms/step - loss: 0.0683 - categorical_accuracy: 0.9487 - val_loss: 0.0522 - val_categorical_accuracy: 0.9930

Epoch 00001: val_loss improved from inf to 0.05222, saving model to weights_base_10000_250_0.1.hdf5
Epoch 2/4
114890/114890 [==============================] - 1664s 14ms/step - loss: 0.0477 - categorical_accuracy: 0.9809 - val_loss: 0.0499 - val_categorical_accuracy: 0.9930

Epoch 00002: val_loss improved from 0.05222 to 0.04993, saving model to weights_base_10000_250_0.1.hdf5
Epoch 3/4
114890/114890 [==============================] - 1671s 15ms/step - loss: 0.0422 - categorical_accuracy: 0.9520 - val_loss: 0.0499 - val_categorical_accuracy: 0.9930

Epoch 00003: val_loss did not improve
Epoch 4/4
114890/114890 [==================

[]
[array([[-0.03326945, -0.03104633, -0.0160776 , ..., -0.02599823,
         0.0395217 , -0.0412238 ],
       [-0.01143156,  0.04759712,  0.00566554, ..., -0.00271648,
        -0.03129635, -0.02996721],
       [ 0.04391935, -0.0277747 , -0.03758749, ..., -0.03850917,
         0.01456382, -0.04619063],
       ...,
       [-0.01271094, -0.04688478, -0.01596321, ...,  0.03239927,
        -0.0429408 ,  0.00881263],
       [ 0.02129759,  0.02610108, -0.02143035, ..., -0.00189376,
        -0.0146777 , -0.00344218],
       [-0.01640928, -0.04305264, -0.0156445 , ..., -0.01497052,
         0.03609885, -0.04355785]], dtype=float32)]
[array([[ 0.08282873,  0.00014524, -0.11439466, ..., -0.02569608,
        -0.13307238, -0.08856292],
       [-0.13233139,  0.03258061, -0.05971719, ...,  0.04307289,
        -0.07603861, -0.12951237],
       [ 0.10117981, -0.11703214, -0.09279917, ..., -0.1214847 ,
         0.01854083, -0.04284104],
       ...,
       [-0.06357735, -0.12858929, -0.09521776, ...,  0

In [19]:
model.Bidirectional.get_weights()

AttributeError: 'Model' object has no attribute 'Bidirectional'

In [10]:
#model=get_model(maxlen=100,max_features=20000,dropout=0.1,embed_size=128)

In [14]:
from keras.models import load_model 

# Returns a compiled model identical to the previous one
#model = load_model("model_name.hdf5")


In [10]:
y_predict = model.predict(x_test_pad)

In [14]:
from keras import metrics

In [22]:
print(y_test.shape)
print(y_predict.shape)

(31915, 6)
(31915, 6)


In [31]:
loaded_model=model

In [34]:
#binary_accuracy(y_true, y_pred)
# evaluate loaded model on test data
loaded_model.compile(loss='binary_crossentropy',optimizer='adam',metrics=[categorical_accuracy])
score = loaded_model.evaluate(x_test_pad, y_test, verbose=1)
print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))

31915/31915 [==============================] - 109s 3ms/step
categorical_accuracy: 99.24%


In [47]:
model_dict={'bidirectional_lstm_globMP_relu_sigmoid_maxfeat20000_maxlen150_dropout0.1':[150,0.1],\
'bidirectional_lstm_globMP_relu_sigmoid_maxfeat20000_maxlen150_dropout0.2':[150,0.2],\
'bidirectional_lstm_globMP_relu_sigmoid_maxfeat20000_maxlen250_dropout0.1':[250,0.1],\
'bidirectional_lstm_globMP_relu_sigmoid_maxfeat20000_maxlen250_dropout0.2':[250,0.2]}

In [49]:
from keras.models import load_model 

# Returns a compiled model identical to the previous one
#model = load_model("model_name.hdf5")

tokenizer = text.Tokenizer(num_words=20000)
#tokenize based on comment database
tokenizer.fit_on_texts(list(train_all["comment_text"].fillna("nada").values)) #fit on all comment_text
#create tokenized comments
#list_tokenized_train = tokenizer.texts_to_sequences(x_train)  #training (80% of train.csv)
list_tokenized_test = tokenizer.texts_to_sequences(x_test)  #testing (20% of train.csv)

for file in model_dict.keys():
    
    maxlen=model_dict[file][0]
    dropout=model_dict[file][1]
    
    print('-'*50,'\nModel with maxlen of {} and dropout of {}'.format(maxlen,dropout))
    
    print('padding maxlen=',maxlen)
    #x_train_pad = sequence.pad_sequences(list_tokenized_train, maxlen=model_dict[file])
    x_test_pad = sequence.pad_sequences(list_tokenized_test, maxlen=maxlen)

    print('loading model')
    loaded_model=load_model(file)
    
    print('compiling model')
    loaded_model.compile(loss='binary_crossentropy',optimizer='adam',metrics=[categorical_accuracy])
    print('evaluating model')
    score = loaded_model.evaluate(x_test_pad, y_test, verbose=1)
    print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))
    print('-'*50,'\n')
    

-------------------------------------------------- 
Model with maxlen of 150 and dropout of 0.1
padding maxlen= 150
loading model
compiling model
evaluating model
31915/31915 [==============================] - 67s 2ms/step
categorical_accuracy: 99.11%
-------------------------------------------------- 

-------------------------------------------------- 
Model with maxlen of 150 and dropout of 0.2
padding maxlen= 150
loading model
compiling model
evaluating model
31915/31915 [==============================] - 84s 3ms/step
categorical_accuracy: 99.26%
-------------------------------------------------- 

-------------------------------------------------- 
Model with maxlen of 250 and dropout of 0.1
padding maxlen= 250
loading model
compiling model
evaluating model
31915/31915 [==============================] - 127s 4ms/step
categorical_accuracy: 97.69%
-------------------------------------------------- 

-------------------------------------------------- 
Model with maxlen of 250 and dro

In [53]:
# import data
print('Importing data...')
train_all=pd.read_csv('train.csv')

# 80/20 split
train_80, test_20 = model_selection.train_test_split(train_all,test_size=0.2)

#preprocessing

print('Splitting data...')

#training data
x_train = train_80['comment_text'].fillna("nada").values
y_train = train_80[['toxic','severe_toxic','obscene','threat','insult','identity_hate']].values

#testing validation data (not for training model, just validation for )

x_test = test_20['comment_text'].fillna("nada").values
y_test = test_20[['toxic','severe_toxic','obscene','threat','insult','identity_hate']].values

# check lengths

print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')



# MODEL PARAMETERS

max_feature_list=[20000,10000]
EPOCHS=4
maxlen_list=[150,250] 
dropout_list=[0.1,0.2]

#TOKENIZE 

tokenizer = text.Tokenizer(num_words=max_feature_list[1])
tokenizer.fit_on_texts(list(train_all["comment_text"].fillna("nada").values)) #fit on all comment_text

#create tokenized comments
list_tokenized_train = tokenizer.texts_to_sequences(x_train)  #training (80% of train.csv)
list_tokenized_test = tokenizer.texts_to_sequences(x_test)  #testing (20% of train.csv)



for max_feature in [max_feature_list[1]]:
    
    for maxlen in [maxlen_list[0]]:
        
        for dropout in [dropout_list[0]]:
            
            print('padding maxlen=',maxlen)
            x_train_pad = sequence.pad_sequences(list_tokenized_train, maxlen=maxlen)
            x_test_pad = sequence.pad_sequences(list_tokenized_test, maxlen=maxlen)
            
            print('building model')
            model=get_model(maxlen=maxlen,max_features=max_feature,dropout=dropout,embed_size=128)
            
            batch_size = 32


            weight_file_path="weights_base_{}_{}_{}.hdf5".format(max_feature,maxlen,dropout)
            model_file_path='bidirectional_lstm_globMP_relu_sigmoid_maxfeat{}_maxlen{}_dropout{}'.format(max_feature,maxlen,dropout)

            checkpoint = ModelCheckpoint(weight_file_path, monitor='val_loss', verbose=1, save_best_only=True, mode='min')

            early = EarlyStopping(monitor="val_loss", mode="min", patience=20)

            callbacks_list = [checkpoint, early] #early
            
            #fit model
            print('fitting model with max_features={}, maxlen = {}, and dropout = {}'.format(max_feature,maxlen,dropout))
            model.fit(x_train_pad, y_train, batch_size=batch_size, epochs=EPOCHS, validation_split=0.1, callbacks=callbacks_list)
            
            #save model
            model.save(model_file_path)


Importing data...
Splitting data...
127656 train sequences
31915 test sequences
padding maxlen= 150
building model
fitting model with max_features=10000, maxlen = 150, and dropout = 0.1
Train on 114890 samples, validate on 12766 samples
Epoch 1/4
114890/114890 [==============================] - 1488s 13ms/step - loss: 0.0701 - categorical_accuracy: 0.9329 - val_loss: 0.0488 - val_categorical_accuracy: 0.9939

Epoch 00001: val_loss improved from inf to 0.04875, saving model to weights_base_10000_150_0.1.hdf5
Epoch 2/4
114890/114890 [==============================] - 1410s 12ms/step - loss: 0.0482 - categorical_accuracy: 0.9766 - val_loss: 0.0470 - val_categorical_accuracy: 0.9934

Epoch 00002: val_loss improved from 0.04875 to 0.04698, saving model to weights_base_10000_150_0.1.hdf5
Epoch 3/4
114890/114890 [==============================] - 1154s 10ms/step - loss: 0.0426 - categorical_accuracy: 0.9715 - val_loss: 0.0461 - val_categorical_accuracy: 0.9934

Epoch 00003: val_loss improved 

## Evaluate model and predictions

In [54]:
loaded_model=model

x_test_pad = sequence.pad_sequences(list_tokenized_test, maxlen=150)

print('compiling model')
loaded_model.compile(loss='binary_crossentropy',optimizer='adam',metrics=[categorical_accuracy])
print('evaluating model')
score = loaded_model.evaluate(x_test_pad, y_test, verbose=1)
print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))
print('-'*50,'\n')

compiling model
evaluating model
31915/31915 [==============================] - 78s 2ms/step
categorical_accuracy: 99.32%
-------------------------------------------------- 



In [97]:
y_predict=model.predict(x_test_pad)

In [98]:
y_predict=pd.DataFrame(y_predict)
y_test=pd.DataFrame(y_test)

y_predict.columns=['toxic','severe_toxic','obscene','threat','insult','identity_hate']
y_test.columns=['toxic','severe_toxic','obscene','threat','insult','identity_hate']


# get predictions with decision boundary = 0.5
for c in ["toxic", "severe_toxic", "obscene", "threat", "insult", "identity_hate"]:

    y_predict[c]=y_predict[c].map(lambda x: 1 if x >=0.5 else 0)
    
# subract predictions from true labels to get type I (= -1) and type II (=1)
y_error=y_test-y_predict



In [99]:
y_error.head(20)

toxic  severe_toxic  obscene  threat  insult  identity_hate
0       0             0        0       0       0              0
1       0             0        0       0       0              0
2       0             0        0       0       0              0
3       0             0        0       0       0              0
4       0             0        0       0       0              0
5       0             0        0       0       0              0
6       0             0        0       0       0              0
7       0             0        0       0       0              0
8       0             0        0       0       0              0
9       0             0        0       0       0              0
10      0             0        0       0       0              0
11      0             0        0       0       0              0
12      0             0        0       0       0              0
13      0             0        0       0       0              0
14      0             0        0       0       0              0
15      0             0        0       0       0              0
16      0             0        0       0       0              0
17      0             0        0       0       0              0
18      1             0        0       0       0              0
19      0             0        0       0       0              0

In [100]:
test_20['comment_text'].iloc[18]

'"\n\nMay Allah (swt) either give you punishment or hidiyaat for spreading falsehood and keeping other Muslim editors from making the article neutral - Insh\'Allah.  —Preceding unsigned comment added by 89.108.24.87     —Preceding unsigned comment added by 213.146.172.146   \n\nYour comment on your edit proves you are ignorant of Islam.  —Preceding unsigned comment added by 213.146.172.146   "'

In [130]:
tot=y_error.shape[0]
for c in ["toxic", "severe_toxic", "obscene", "threat", "insult", "identity_hate"]:
    print(c)
    print(y_error[c].value_counts())
    type1=0
    type2=0
    
    try:
        type1=y_error[c].value_counts()[-1]
    except:
        pass
    
    try:
        type2=y_error[c].value_counts()[1]
    except:
        pass
    print('\n Type I (flagged when no flag exists):','{:02.2f}%'.format(type1*100/tot))
    print('Type II (not flagged when flag should exist):',"{:02.2f}%".format(type2*100/tot))
    print('\n \n')

toxic
 0    30690
 1      992
-1      233
Name: toxic, dtype: int64

 Type I (flagged when no flag exists): 0.73%
Type II (not flagged when flag should exist): 3.11%

 

severe_toxic
 0    31611
 1      187
-1      117
Name: severe_toxic, dtype: int64

 Type I (flagged when no flag exists): 0.37%
Type II (not flagged when flag should exist): 0.59%

 

obscene
 0    31292
 1      384
-1      239
Name: obscene, dtype: int64

 Type I (flagged when no flag exists): 0.75%
Type II (not flagged when flag should exist): 1.20%

 

threat
0    31829
1       86
Name: threat, dtype: int64

 Type I (flagged when no flag exists): 0.00%
Type II (not flagged when flag should exist): 0.27%

 

insult
 0    31060
 1      532
-1      323
Name: insult, dtype: int64

 Type I (flagged when no flag exists): 1.01%
Type II (not flagged when flag should exist): 1.67%

 

identity_hate
 0    31675
 1      161
-1       79
Name: identity_hate, dtype: int64

 Type I (flagged when no flag exists): 0.25%
Type II (not

In [138]:
test_20.shape

(31915, 8)

In [145]:
model.weights[7]

<tf.Variable 'dense_11/kernel:0' shape=(100, 50) dtype=float32_ref>


In [152]:

wrong=0
for index, row in y_error.iterrows():
    if y_error.iloc[index].value_counts()[0]!=6:
        wrong+=1
        
print(wrong/y_error.shape[0])

0.07773774087419709


In [151]:
y_error.iloc[1].value_counts()[0]

6

In [153]:
sum([992,233,187,117,384,239,86,532,323,161,79])

3333

In [155]:
0.07773774087419709/6

0.012956290145699515

## NOW LETS AUTOMATE MAX FEATURES 10000

In [158]:
max_feature_list=[20000,10000]
EPOCHS=4
maxlen_list=[150,250] 
dropout_list=[0.1,0.2]

In [166]:
# tokenize

tokenizer = text.Tokenizer(num_words=max_feature_list[1])

tokenizer.fit_on_texts(list(train_all["comment_text"].fillna("nada").values)) #fit on all comment_text

#create tokenized comments
list_tokenized_train = tokenizer.texts_to_sequences(x_train)  #training (80% of train.csv)
list_tokenized_test = tokenizer.texts_to_sequences(x_test)  #testing (20% of train.csv)

In [167]:
max_feature_list=[20000,10000]
maxlen_list=[150,250] 
dropout_list=[0.1,0.2]
EPOCHS=4

for max_feature in [max_feature_list[1]]:
    
    for maxlen in maxlen_list:
        
        for dropout in dropout_list:
            
            if (maxlen,dropout) !=(150,0.1):  # already did 150,0.1
                print('padding maxlen=',maxlen)
                x_train_pad = sequence.pad_sequences(list_tokenized_train, maxlen=maxlen)
                x_test_pad = sequence.pad_sequences(list_tokenized_test, maxlen=maxlen)

                print('building model')
                model=get_model(maxlen=maxlen,max_features=max_feature,dropout=dropout,embed_size=128)

                batch_size = 32


                weight_file_path="weights_base_{}_{}_{}.hdf5".format(max_feature,maxlen,dropout)
                model_file_path='bidirectional_lstm_globMP_relu_sigmoid_maxfeat{}_maxlen{}_dropout{}'.format(max_feature,maxlen,dropout)

                checkpoint = ModelCheckpoint(weight_file_path, monitor='val_loss', verbose=1, save_best_only=True, mode='min')

                early = EarlyStopping(monitor="val_loss", mode="min", patience=20)

                callbacks_list = [checkpoint, early] #early

                #fit model
                print('fitting model with max_features={}, maxlen = {}, and dropout = {}'.format(max_feature,maxlen,dropout))
                model.fit(x_train_pad, y_train, batch_size=batch_size, epochs=EPOCHS, validation_split=0.1, callbacks=callbacks_list)

                #save model
                model.save(model_file_path)

padding maxlen= 150
building model
fitting model with max_features=10000, maxlen = 150, and dropout = 0.2
Train on 114890 samples, validate on 12766 samples
Epoch 1/4
114890/114890 [==============================] - 1208s 11ms/step - loss: 0.0695 - categorical_accuracy: 0.9150 - val_loss: 0.0489 - val_categorical_accuracy: 0.9940

Epoch 00001: val_loss improved from inf to 0.04887, saving model to weights_base_10000_150_0.2.hdf5
Epoch 2/4
114890/114890 [==============================] - 1168s 10ms/step - loss: 0.0488 - categorical_accuracy: 0.9832 - val_loss: 0.0475 - val_categorical_accuracy: 0.9940

Epoch 00002: val_loss improved from 0.04887 to 0.04755, saving model to weights_base_10000_150_0.2.hdf5
Epoch 3/4
114890/114890 [==============================] - 1158s 10ms/step - loss: 0.0437 - categorical_accuracy: 0.9822 - val_loss: 0.0469 - val_categorical_accuracy: 0.9938

Epoch 00003: val_loss improved from 0.04755 to 0.04693, saving model to weights_base_10000_150_0.2.hdf5
Epoch 4

In [161]:
max_feature_list=[20000,10000]
maxlen_list=[150,250] 
dropout_list=[0.1,0.2]
EPOCHS=4

for max_feature in [max_feature_list[1]]:
    
    for maxlen in maxlen_list:
        
        for dropout in dropout_list:
            
            if (maxlen,dropout) !=(150,0.1):
                
                print(max_feature,maxlen,dropout)

10000 150 0.2
10000 250 0.1
10000 250 0.2


## Evaluate on test data

In [168]:
model_dict={'bidirectional_lstm_globMP_relu_sigmoid_maxfeat10000_maxlen150_dropout0.1':[150,0.1],\
'bidirectional_lstm_globMP_relu_sigmoid_maxfeat10000_maxlen150_dropout0.2':[150,0.2],\
'bidirectional_lstm_globMP_relu_sigmoid_maxfeat10000_maxlen250_dropout0.1':[250,0.1],\
'bidirectional_lstm_globMP_relu_sigmoid_maxfeat10000_maxlen250_dropout0.2':[250,0.2]}

In [169]:
from keras.models import load_model 

# Returns a compiled model identical to the previous one
#model = load_model("model_name.hdf5")

tokenizer = text.Tokenizer(num_words=10000)
#tokenize based on comment database
tokenizer.fit_on_texts(list(train_all["comment_text"].fillna("nada").values)) #fit on all comment_text
#create tokenized comments
#list_tokenized_train = tokenizer.texts_to_sequences(x_train)  #training (80% of train.csv)
list_tokenized_test = tokenizer.texts_to_sequences(x_test)  #testing (20% of train.csv)

for file in model_dict.keys():
    
    maxlen=model_dict[file][0]
    dropout=model_dict[file][1]
    
    print('-'*50,'\nModel with max_features 10000, maxlen of {}, and dropout of {}'.format(maxlen,dropout))
    
    print('padding maxlen=',maxlen)
    #x_train_pad = sequence.pad_sequences(list_tokenized_train, maxlen=model_dict[file])
    x_test_pad = sequence.pad_sequences(list_tokenized_test, maxlen=maxlen)

    print('loading model')
    loaded_model=load_model(file)
    
    print('compiling model')
    loaded_model.compile(loss='binary_crossentropy',optimizer='adam',metrics=[categorical_accuracy])
    print('evaluating model')
    score = loaded_model.evaluate(x_test_pad, y_test, verbose=1)
    print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))
    print('-'*50,'\n')

-------------------------------------------------- 
Model with max_features 10000, maxlen of 150, and dropout of 0.1
padding maxlen= 150
loading model
compiling model
evaluating model
31915/31915 [==============================] - 84s 3ms/step
categorical_accuracy: 99.27%
-------------------------------------------------- 

-------------------------------------------------- 
Model with max_features 10000, maxlen of 150, and dropout of 0.2
padding maxlen= 150
loading model
compiling model
evaluating model
31915/31915 [==============================] - 80s 3ms/step
categorical_accuracy: 99.35%
-------------------------------------------------- 

-------------------------------------------------- 
Model with max_features 10000, maxlen of 250, and dropout of 0.1
padding maxlen= 250
loading model
compiling model
evaluating model
31915/31915 [==============================] - 126s 4ms/step
categorical_accuracy: 99.35%
-------------------------------------------------- 

---------------------

In [170]:
model.layers

In [173]:
print('loading model')
loaded_model=load_model('bidirectional_lstm_globMP_relu_sigmoid_maxfeat10000_maxlen250_dropout0.2')

print('compiling model')
loaded_model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['binary_accuracy','categorical_accuracy'])
print('evaluating model')
score = loaded_model.evaluate(x_test_pad, y_test, verbose=1)
print("%s: %.2f%%" % (loaded_model.metrics_names[:], score[1]*100))
print('-'*50,'\n')

loading model
compiling model
evaluating model
31915/31915 [==============================] - 129s 4ms/step
binary_accuracy: 98.31%
-------------------------------------------------- 



In [181]:
print("%s: %.2f%%" % (loaded_model.metrics_names[0], score[0]*100))
print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))
print("%s: %.2f%%" % (loaded_model.metrics_names[2], score[2]*100))
print("%s: %.2f%%" % (loaded_model.metrics_names[3], score[3]*100))

loss: 4.92%
binary_accuracy: 98.31%
categorical_accuracy: 99.15%
sparse_categorical_accuracy: 89.87%


## AUTOMATE ACCURACY TEST ON NON-MAXPOOLED

In [35]:
test_model_dict={'bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen150_dropout0.1':[10000,150],\
                 'bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen250_dropout0.1':[10000,250],\
                'bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen150_dropout0.1':[20000,150],\
                'bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen250_dropout0.1':[20000,250],}

In [ ]:
from keras.models import load_model

In [ ]:
tokenizer = text.Tokenizer(num_words=max_features)
tokenizer.fit_on_texts(list(list_sentences_train))
list_tokenized_train = tokenizer.texts_to_sequences(list_sentences_train)
#list_tokenized_test = tokenizer.texts_to_sequences(list_sentences_test)
X_t = sequence.pad_sequences(list_tokenized_train, maxlen=maxlen)

In [38]:
train = pd.read_csv("train.csv")
list_sentences_train_all = train["comment_text"].fillna("nada").values
list_sentences_train_all.shape

(159571,)

In [144]:
x_test=pd.read_csv('x_test.csv',header=-1)
x_test.head()

0
0  He just MIGHT have been telling the truth but ...
1  "\nNo, it's no, and I know it. I also know (an...
2  Kurt Cobain\nwhy do you think you own the Kurt...
3  Edit request \n\nThank you for , can you do th...
4  AlphaChump \n\nAttempting to erase history is ...

In [123]:
#build before running

performance_dict=dict()

for key in test_model_dict.keys():
    performance_dict[key]={'total_toxic':0,'toxic_falsepos':0,'toxic_falseneg':0,\
                           'toxic_falsepos_comment':[],'toxic_falseneg_comment':[]}

In [145]:
list_classes = ["toxic", "severe_toxic", "obscene", "threat", "insult", "identity_hate"]

test_model_dict={'bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen150_dropout0.1':[10000,150],\
                 'bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen250_dropout0.1':[10000,250],\
                'bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen150_dropout0.1':[20000,150],\
                'bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen250_dropout0.1':[20000,250],}

performance_dict=dict()

for key in test_model_dict.keys():
    performance_dict[key]={'total_toxic':0,'toxic_falsepos':0,'toxic_falseneg':0,\
                           'toxic_falsepos_comment':[],'toxic_falseneg_comment':[]}

#create correct classification array keep as DataFrame
gold=pd.read_csv('gold.csv')
gold.columns=list_classes

#load test data and convert to np.array
x_test=pd.read_csv('x_test.csv',header=-1)
x_test=x_test[0].fillna("nada").values

for key in test_model_dict.keys():
    
    print(key)
    # load model
    print('loading model')
    test_model=load_model(key)
    
    # tokenize x_test
    print('tokenizing')
    max_features=test_model_dict[key][0] #max features for model
    maxlen=test_model_dict[key][1] #max sequence length for model
    
    tokenizer = text.Tokenizer(num_words=max_features)    # set up tokenizer
    tokenizer.fit_on_texts(list(list_sentences_train_all)) # fit tokenizer to all in train.csv
    tokenized_x_test = tokenizer.texts_to_sequences(x_test) # tokenize x_test
    test_sequence=sequence.pad_sequences(tokenized_x_test, maxlen=maxlen) #create the test sequence
    
    # 
    print('compiling model')
    test_model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=[categorical_accuracy])
    
    print('evaluating model')
    score = test_model.evaluate(test_sequence, y_test, verbose=1)
    
    print("%s: %.2f%%" % (test_model.metrics_names[1], score[1]*100))
    
    print('creating prediction')
    predict=test_model.predict(test_sequence, verbose=1)
    
    print('test predictions on "toxic" category')
    predict_df=pd.DataFrame(predict)
    predict_df.columns=list_classes
    
    predict_df.to_csv('prediction_'+key+'.csv')
    
    for c in predict_df.columns:
         predict_df[c]=predict_df[c].map(lambda x: 1 if x >=0.5 else 0)

    err_df=gold['toxic']-predict_df['toxic']  #val of -1 is false positive Type I, value of +1 is false negative Type II
    
    performance_dict[key]['total_toxic']=sum(gold['toxic'])
    performance_dict[key]['toxic_falsepos']=err_df.value_counts()[-1]
    performance_dict[key]['toxic_falseneg']=err_df.value_counts()[1]
    
    print('total true toxic',sum(gold['toxic']))
    #false positives
    print('total false positive (incorrectly flagged as toxic) Type I',err_df.value_counts()[-1])
    print('false positive rate', '{:02.3f}'.format(100*err_df.value_counts()[-1]/sum(gold['toxic'])),'%')
    #false negatives
    print('total false negative (omitted flagging as toxic) Type II',err_df.value_counts()[1])
    print('false negative rate', '{:02.3f}'.format(100*err_df.value_counts()[1]/sum(gold['toxic'])),'%')
    
    
    print('-'*50,'\n')
     

bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen150_dropout0.1
loading model
tokenizing
compiling model
evaluating model
31915/31915 [==============================] - 77s 2ms/step
categorical_accuracy: 99.39%
creating prediction
31915/31915 [==============================] - 75s 2ms/step
test predictions on "toxic" category
total true toxic 2933
total false positive (incorrectly flagged as toxic) Type I 744
false positive rate 25.367 %
total false negative (omitted flagging as toxic) Type II 670
false negative rate 22.844 %
-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen250_dropout0.1
loading model
tokenizing
compiling model
evaluating model
31915/31915 [==============================] - 115s 4ms/step
categorical_accuracy: 98.50%
creating prediction
31915/31915 [==============================] - 114s 4ms/step
test predictions on "toxic" category
total true toxic 2933
total false positive (incorrectly flagged as toxic) Type I 

bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen150_dropout0.1
loading model
tokenizing
compiling model
evaluating model
31915/31915 [==============================] - 77s 2ms/step

categorical_accuracy: 99.39%

creating prediction

31915/31915 [==============================] - 75s 2ms/step

test predictions on "toxic" category

total true toxic 2933

total false positive (incorrectly flagged as toxic) Type I 744

false positive rate 25.367 %

total false negative (omitted flagging as toxic) Type II 670

false negative rate 22.844 %


-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen250_dropout0.1

loading model

tokenizing

compiling model

evaluating model

31915/31915 [==============================] - 115s 4ms/step

categorical_accuracy: 98.50%

creating prediction

31915/31915 [==============================] - 114s 4ms/step

test predictions on "toxic" category

total true toxic 2933

total false positive (incorrectly flagged as toxic) Type I 616

false positive rate 21.002 %

total false negative (omitted flagging as toxic) Type II 669

false negative rate 22.809 %

-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen150_dropout0.1

loading model

tokenizing

compiling model

evaluating model

31915/31915 [==============================] - 86s 3ms/step

categorical_accuracy: 99.36%

creating prediction

31915/31915 [==============================] - 78s 2ms/step

test predictions on "toxic" category

total true toxic 2933

total false positive (incorrectly flagged as toxic) Type I 222

false positive rate 7.569 %

total false negative (omitted flagging as toxic) Type II 670

false negative rate 22.844 %

-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen250_dropout0.1

loading model

tokenizing

compiling model

evaluating model

31915/31915 [==============================] - 112s 4ms/step

categorical_accuracy: 99.40%

creating prediction

31915/31915 [==============================] - 106s 3ms/step

test predictions on "toxic" category

total true toxic 2933

total false positive (incorrectly flagged as toxic) Type I 432

false positive rate 14.729 %

total false negative (omitted flagging as toxic) Type II 326

false negative rate 11.115 %

-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen150_dropout0.1
loading model
tokenizing
compiling model
evaluating model
31915/31915 [==============================] - 73s 2ms/step
categorical_accuracy: 99.39%
-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat10000_maxlen250_dropout0.1
loading model
tokenizing
compiling model
evaluating model
31915/31915 [==============================] - 113s 4ms/step
categorical_accuracy: 98.50%
-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen150_dropout0.1
loading model
tokenizing
compiling model
evaluating model
31915/31915 [==============================] - 74s 2ms/step
categorical_accuracy: 99.36%
-------------------------------------------------- 

bidirectional_lstm_relu_sigmoid_maxfeat20000_maxlen250_dropout0.1
loading model
tokenizing
compiling model
evaluating model
31915/31915 [==============================] - 116s 4ms/step
categorical_accuracy: 99.40%
-------------------------------------------------- 

In [50]:
gold=pd.DataFrame(y_test)
gold[0][:10]

0    0
1    0
2    0
3    0
4    0
5    0
6    1
7    0
8    0
9    0
Name: 0, dtype: int64

In [51]:
predict=test_model.predict(test_sequence, verbose=1)

31915/31915 [==============================] - 102s 3ms/step


In [60]:
predict_df=pd.DataFrame(predict)

for c in predict_df.columns:

    predict_df[c]=predict_df[c].map(lambda x: 1 if x >=0.5 else 0)

err_df=gold[0]-predict_df[0]

In [73]:
print('{:02.3f}'.format(100*err_df.value_counts()[-1]/sum(gold[0])))

14.729


In [139]:
x_test

array([['"\nNo, it\'s no, and I know it. I also know (and understand) about WP:FILMPLOT, which is something you appear not to have read. -  (^ • @) "'],
       ['Kurt Cobain\nwhy do you think you own the Kurt Cobain page? I am not trying to attack you, I am just wondering why.'],
       ['Edit request \n\nThank you for , can you do the same for  please?'],
       ...,
       ["Dictionaries\n\nHere are some dictionaries that do not define the word as always offensive: Oxford English Dictionary Online[]\n'2. A native of China.'\nMerriam-Webster Online Dictionary[]\n'often offensive  a native of China  CHINESE '\nI can find more if needed. I think that these counter-examples can allow the word 'many' to be inserted in front of 'dictionaries' in saying that the word is offensive."],
       ["Provided that it's removed entirely to here, on your head be it! ;)"],
       ['fucking me hard because i am an e-geek']], dtype=object)

In [95]:
#create correct classification array


gold=pd.read_csv('gold.csv').values
gold[:10]

#load test data and convert to np.array
#x_test=pd.read_csv('x_test.csv')
#x_test=x_test.fillna("nada").values
#x_text[:2]

array([[0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [1, 0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0]])

In [100]:
for key in test_model_dict.keys():
    performance_dict[key]={'total_toxic':0,'toxic_falsepos':0,'toxic_falseneg':0,\
                           'toxic_falsepos_comment':[],'toxic_falseneg_comment':[]}

In [115]:

dummy= pd.DataFrame(gold['toxic']-gold['threat'])
dummy.columns=['error']

dummy.shape

(31915, 1)

In [140]:
x_test.head()

AttributeError: 'numpy.ndarray' object has no attribute 'head'

error    0
0    0.0  NaN
1    0.0  NaN
2    0.0  NaN
3    0.0  NaN
4    0.0  NaN